Added partial support for wpa_supplicant, hostapd, and cjose:

julek-wolfssl opened the pull request:
- Moved `SetECKeyInternal` and `SetECKeyExternal` to `internal.h` to allow usage outside of `ssl.c`
- Added `asn1t.h`
- Implemented the `IMPLEMENT_ASN1_FUNCTIONS` macro for a small subset of ASN1 tags
-- So far only `X509_ALGOR` and `ASN1_BIT_STRING` are supported
- Implemented `BN_mod_add` function
- Allow for setting of `EC_KEY` export form through EC_KEY_set_conv_form
- Implemented `i2o_ECPublicKey`
- Implemented `EC_POINT_copy`
- Implemented deriving DH and ECDH keys in `EVP_PKEY_CTX`. Functions added:
-- `EVP_PKEY_derive_init`
-- `EVP_PKEY_derive_set_peer`
-- `EVP_PKEY_derive`
- Implemented `EVP_PKEY_get0_DH`
- Implemented `X509_ALGOR_new`
- Implemented `X509_ALGOR_free`
- Implemented `X509_ALGOR_set0`
- Implemented `X509_PUBKEY_new`
- Implemented `X509_PUBKEY_free`
- Implemented `X509_PUBKEY_set`
- Implemented `RSA_padding_add_PKCS1_PSS`
- Implemented `RSA_verify_PKCS1_PSS`
- Changed second parameter of `wolfSSL_d2i_PUBKEY` to be constant
- Corrected long names in `asn.h`
- Added `wc_ecc_get_generator` as a way to get the generator point of a curve
- Added `wc_ecc_export_point_der_ex` to export an ECC point in compressed or uncompressed format with one API
- Added `wc_ecc_export_point_der_compressed` to export a point in an `ecc_point` structure in compressed DER format
- Added 'wc_RsaSSL_Verify_ex` which adds the option to choose a padding type
- Added `wc_RsaPad_ex` and `wc_RsaUnPad_ex` to `rsa.h` as `WOLFSSL_LOCAL` functions
- `CopyDecodedToX509` now fills `x509->key` and `x509->algor` when populating x509
- `wolfSSL_EVP_CipherInit` now uses `wc_AesGcmSetExtIV` to set the IV so that it is copied to `ctx->iv` by `wolfSSL_StoreExternalIV`
- Added error checking to `wolfSSL_EVP_PKEY_get_der`
- `wolfSSL_X509_ALGOR_get0` now attempts to return something in all parameters
- Refactored `wolfSSL_EC_KEY_new` to use `*_new` functions when available
- Added `setupPoint` to set the internal point if not yet set
- Always set external point in `wolfSSL_ECPoint_d2i`
- Added compressed point support to `wolfSSL_EC_POINT_point2oct`
- Fix `wolfSSL_EC_POINT_mul` so that it will calculate the full `generator * n + q * m` then OpenSSL does
- Added `WOLFSSL_RSA_GetRNG` helper function to get a `WC_RNG` from `WOLFSSL_RSA`
- Correct short names in `wolfssl_object_info`
- Added all currently supported curves to `wolfssl_object_info`
- Added `oidCurveType` to `oid2nid`
- Add more padding types to `wolfSSL_RSA_public_decrypt`
- Fix `keysize` in `wc_ecc_import_point_der`
- Added tests for new additions

julek-wolfssl committed:
Added partial support for wpa_supplicant, hostapd, and cjose:

- Moved `SetECKeyInternal` and `SetECKeyExternal` to `internal.h` to allow usage outside of `ssl.c`
- Added `asn1t.h`
- Implemented the `IMPLEMENT_ASN1_FUNCTIONS` macro for a small subset of ASN1 tags
-- So far only `X509_ALGOR` and `ASN1_BIT_STRING` are supported
- Implemented `BN_mod_add` function
- Allow for setting of `EC_KEY` export form through EC_KEY_set_conv_form
- Implemented `i2o_ECPublicKey`
- Implemented `EC_POINT_copy`
- Implemented deriving DH and ECDH keys in `EVP_PKEY_CTX`. Functions added:
-- `EVP_PKEY_derive_init`
-- `EVP_PKEY_derive_set_peer`
-- `EVP_PKEY_derive`
- Implemented `EVP_PKEY_get0_DH`
- Implemented `X509_ALGOR_new`
- Implemented `X509_ALGOR_free`
- Implemented `X509_ALGOR_set0`
- Implemented `X509_PUBKEY_new`
- Implemented `X509_PUBKEY_free`
- Implemented `X509_PUBKEY_set`
- Implemented `RSA_padding_add_PKCS1_PSS`
- Implemented `RSA_verify_PKCS1_PSS`
- Changed second parameter of `wolfSSL_d2i_PUBKEY` to be constant
- Corrected long names in `asn.h`
- Added `wc_ecc_get_generator` as a way to get the generator point of a curve
- Added `wc_ecc_export_point_der_ex` to export an ECC point in compressed or uncompressed format with one API
- Added `wc_ecc_export_point_der_compressed` to export a point in an `ecc_point` structure in compressed DER format
- Added 'wc_RsaSSL_Verify_ex` which adds the option to choose a padding type
- Added `wc_RsaPad_ex` and `wc_RsaUnPad_ex` to `rsa.h` as `WOLFSSL_LOCAL` functions
- `CopyDecodedToX509` now fills `x509->key` and `x509->algor` when populating x509
- `wolfSSL_EVP_CipherInit` now uses `wc_AesGcmSetExtIV` to set the IV so that it is copied to `ctx->iv` by `wolfSSL_StoreExternalIV`
- Added error checking to `wolfSSL_EVP_PKEY_get_der`
- `wolfSSL_X509_ALGOR_get0` now attempts to return something in all parameters
- Refactored `wolfSSL_EC_KEY_new` to use `*_new` functions when available
- Added `setupPoint` to set the internal point if not yet set
- Always set external point in `wolfSSL_ECPoint_d2i`
- Added compressed point support to `wolfSSL_EC_POINT_point2oct`
- Fix `wolfSSL_EC_POINT_mul` so that it will calculate the full `generator * n + q * m` then OpenSSL does
- Added `WOLFSSL_RSA_GetRNG` helper function to get a `WC_RNG` from `WOLFSSL_RSA`
- Correct short names in `wolfssl_object_info`
- Added all currently supported curves to `wolfssl_object_info`
- Added `oidCurveType` to `oid2nid`
- Add more padding types to `wolfSSL_RSA_public_decrypt`
- Fix `keysize` in `wc_ecc_import_point_der`
- Added tests for new additions

julek-wolfssl committed:
Test fixes

Free `x509->key.pkey` in `FreeX509
Fix type conversions
Fix memory leaks and use of uninitialized memory

julek-wolfssl committed:
Test fixes

Remove redundant macros

julek-wolfssl committed:
Test fixes

Config fixes
Fix windows FIPS

julek-wolfssl committed:
Test fixes

julek-wolfssl committed:
Test fixes

julek-wolfssl committed:
Fix `nid2*` and `*2nid` functions

toddouska commented:
Jenkins please test again

SparkiDev commented on src/ssl.c:
```diff
@@ -30956,6 +31431,210 @@ WOLFSSL_ASN1_INTEGER* wolfSSL_BN_to_ASN1_INTEGER(const WOLFSSL_BIGNUM *bn, WOLFS
+    for (member = template->members, i = 0; i < template->mcount;
+            member++, i++) {
+        switch(member->type) {
+        case WOLFSSL_X509_ALGOR_ASN1:
```
4 more spaces before case statements.

SparkiDev commented on src/ssl.c:
```diff
@@ -30956,6 +31431,210 @@ WOLFSSL_ASN1_INTEGER* wolfSSL_BN_to_ASN1_INTEGER(const WOLFSSL_BIGNUM *bn, WOLFS
+            if (!algor) {
+                goto error;
+            }
+            *(WOLFSSL_X509_ALGOR**)(((byte*)ret) + member->offset) = algor;
```
Why is an offset needed?

julek-wolfssl replied:
To allow for custom user struct's to be used with `IMPLEMENT_ASN1_FUNCTIONS`.

SparkiDev commented on src/ssl.c:
```diff
@@ -36205,72 +36929,112 @@ int wolfSSL_EC_POINT_set_affine_coordinates_GFp(const WOLFSSL_EC_GROUP *group,
 
-#ifndef WOLFSSL_ATECC508A
-/* return code compliant with OpenSSL :
+#if !defined(WOLFSSL_ATECC508A) && defined(ECC_SHAMIR) && !defined(HAVE_SELFTEST)
```
If not ECC_SHAMIR then you can still multiply each separately and add after.

SparkiDev commented on src/ssl.c:
```diff
@@ -38184,6 +38973,246 @@ int wolfSSL_RSA_LoadDer_ex(WOLFSSL_RSA* rsa, const unsigned char* derBuf,
+                                      const unsigned char *mHash,
+                                      const WOLFSSL_EVP_MD *hashAlg, int saltLen)
+{
+    int hLen, emLen, mgf;
```
Is this a copy of the code from rsa.c?

julek-wolfssl replied:
This isn't a copy. I couldn't find a way to only add padding in `rsa.c` so this is the result.

SparkiDev replied:
This code looks borrowed from OpenSSL -rsa_pss.c.
Rewrite.

julek-wolfssl replied:
Rewritten with some refactoring as well.

SparkiDev commented on src/ssl.c:
```diff
@@ -47478,6 +48136,68 @@ int oid2nid(word32 oid, int grp)
+
+    #ifdef HAVE_ECC
+        case oidCurveType:
+            switch (oid) {
```
This looks like a copy of EccNumToNID()???

julek-wolfssl replied:
Its similar. `EccEnumToNID` converts our internal curve ID's (`enum ecc_curve_id` found in `ecc.h`) to NID's while this converts the Object ID's (`enum Ecc_Sum` found in `asn_public.h`) to NID's. Our ID's and the Object ID's don't have the same values so `EccEnumToNID` can't be used.

SparkiDev commented on wolfcrypt/src/asn.c:
```diff
@@ -5564,45 +5564,25 @@ WOLFSSL_API int EccEnumToNID(int n)
-            eccEnum = ecc_sets[i].id;
-            /* Convert enum value in ecc_curve_id to OpenSSL NID */
-            return EccEnumToNID(eccEnum);
+    const WOLFSSL_ObjectInfo *obj_info = wolfssl_object_info;
```
Where is wolfssl_object_info coming from???
The OpenSSL compatibility layer is meant to be a thin layer on top.
Don't mix layers.
Remember we are embedded a lot of the time so we want small data.
Don't make us build in data we don't have to.

julek-wolfssl replied:
`wolfssl_object_info` is found in ssl.c. It is the struct that contains all of the NID information. When compiling with `OPENSSL_EXTRA_X509_SMALL` instead of `OPENSSL_EXTRA` then only information required for certificates in included in the struct. If `NO_CERTS` is defined then it is an empty struct. Should I revert this function to its previous version and modify `wolfSSL_OBJ_sn2nid` to use `wolfssl_object_info` instead of `wc_OBJ_sn2nid`?

SparkiDev replied:
wc_OBJ_sn2nid calls wc_OBJ_sn2nid.
The data needs to be brought down from ssl.c into asn.c not data being referenced from asn.c but in ssl.c.
From what I can tell in a brief look wolfssl_object_info was only compiled with OPENSSL_EXTRA and not OPENSSL_EXTRA_X509_SMALL.
This move has greater implications. There are many other functions that need this data in ssl.c.
This really needs a separate pull request to fix.

Is there any reason we can't leave wc_OBJ_s2nid as it was?

julek-wolfssl replied:
I changed it so that only a subset needed by `asn.c` would be compiled with `OPENSSL_EXTRA_X509_SMALL`. My reasoning for making `wc_OBJ_sn2nid` use `wolfssl_object_info` was to provide easy extensibility. The previous implementation had a very small subset of NID's hard coded in the function which was not enough for `wolfSSL_OBJ_sn2nid`. It is possible to keep `wc_OBJ_sn2nid` as it was and have `wolfSSL_OBJ_sn2nid` use `wolfssl_object_info` directly. I didn't want to introduce "duplicate" code and leave `wc_OBJ_sn2nid` "behind" in terms of correctness and general usability. I understand that we want to keep build size small which is why I stripped down `wolfssl_object_info`. Now I see that I removed `wc_OBJ_sn2nid`'s ability to recognize ECC names as well so those would also have to be included in the small `wolfssl_object_info` to maintain the same functionality.

SparkiDev replied:
Revert the change.
Make a new pull request that fixes this up once this has been merged.

SparkiDev commented on wolfcrypt/src/ecc.c:
```diff
@@ -6300,8 +6301,8 @@ int wc_ecc_import_point_der(byte* in, word32 inLen, const int curve_idx,
-    /* calculate key size based on inLen / 2 */
-    keysize = inLen>>1;
+    /* calculate key size based on inLen / 2 if uncompressed */
+    keysize = compressed ? inLen : inLen>>1;
```
Good fix, but it will break things.

julek-wolfssl replied:
What is the alternative?

SparkiDev replied:
New API.
Need to have an extended API that does things properly with different behaviour.

SparkiDev replied:
Leave it as it is.
It is used for compressed as it is.

julek-wolfssl replied:
What should I call the new API?

SparkiDev replied:
Yep!

julek-wolfssl replied:
Created a new API and preserved the functionality of the old one.

SparkiDev commented on src/ssl.c:
```diff
@@ -34894,12 +35575,9 @@ static int SetECKeyExternal(WOLFSSL_EC_KEY* eckey)
-#ifdef HAVE_ECC
 /* EC_KEY Openssl -> WolfSSL */
-static int SetECKeyInternal(WOLFSSL_EC_KEY* eckey)
+int SetECKeyInternal(WOLFSSL_EC_KEY* eckey)
```
Should still be static as far as I can tell.
Same for SetECKeyExternal.

julek-wolfssl replied:
I use both in `wolfSSL_EVP_PKEY_derive`.

SparkiDev replied:
I would like these functions to be in wolfcrypt/src/evp.c but am checking with Todd.

SparkiDev replied:
Leave as is.

SparkiDev commented on wolfcrypt/src/ecc.c:
```diff
@@ -6471,6 +6485,69 @@ int wc_ecc_export_point_der(const int curve_idx, ecc_point* point, byte* out,
 }
 
 
+/* export point to der */
```
Protect with:
#ifdef HAVE_COMP_KEY

SparkiDev commented on wolfcrypt/src/ecc.c:
```diff
@@ -6857,6 +6933,30 @@ static int ecc_check_pubkey_order(ecc_key* key, ecc_point* pubkey, mp_int* a,
 #endif
 #endif /* !WOLFSSL_ATECC508A && !WOLFSSL_CRYPTOCELL*/
 
+int wc_ecc_get_generator(ecc_point* ecp, int curve_idx)
```
This is not going to be needed very often.
Protect with OPENSSLALL define check.

julek-wolfssl replied:
Protected with `OPENSSL_EXTRA` as it is used in `wolfSSL_EC_POINT_mul`.

SparkiDev commented on wolfssl/wolfcrypt/ecc.h:
```diff
@@ -422,6 +422,7 @@ WOLFSSL_ABI WOLFSSL_API void wc_ecc_key_free(ecc_key*);
 
 /* ECC predefined curve sets  */
 extern const ecc_set_type ecc_sets[];
+extern const size_t ecc_sets_count;
```
ecc_sets has ECC_CURVE_INVALID as last entry.
Why do you need ecc_sets_count?

julek-wolfssl replied:
I used `ecc_sets_count` in `wolfSSL_EC_get_builtin_curves` so that it doesn't have to be computed at each function call.

SparkiDev commented on wolfssl/openssl/asn1.h:
```diff
@@ -24,6 +24,9 @@
 #define WOLFSSL_ASN1_H_
 
+#ifdef WOLFSSL_WPAS
+#include <stddef.h>  /* for offsetof */
```
offsetof is not portable for old embedded compilers.
Sorry, you'll have to find another way to do this.

julek-wolfssl replied:
It is possible to define `offsetof` as `((size_t)((char *)&((st *)0)->m - (char *)0))` would this be fine? An alternative would be to declare a dummy static struct and calculate the offset based on that struct. For now I implemented the second method and it seems to work. Let me know what you think.

SparkiDev replied:
When I looked around this code I can see that the offset is required.
So, defining offsetof when not available is fine.

SparkiDev replied:
Use a dummy static struct (I presume you mean a static const global variable).
I have a vague memory of compilers not liking NULL.

julek-wolfssl commented on configure.ac:
```diff
@@ -554,14 +554,14 @@ AC_ARG_ENABLE([opensslall],
 [ ENABLED_OPENSSLALL=no ]
 )
-if test "$ENABLED_LIBWEBSOCKETS" = "yes" || test "$ENABLED_OPENVPN" = "yes"
+if test "$ENABLED_LIBWEBSOCKETS" = "yes" || test "$ENABLED_OPENVPN" = "yes" || test "$ENABLED_WPAS" = "yes"
```
This is necessary to support DPP. Alternatively we could have it somewhere in the documentation that DPP additionally requires `--enable-opensslall`?

SparkiDev replied:
What APIs are needed for DPP

julek-wolfssl replied:
These are the functions that are missing when not using OPENSSLALL:
EC_POINT_is_on_curve
i2d_ECPrivateKey
i2d_EC_PUBKEY
d2i_ECPrivateKey
ASN1_SEQUENCE
ASN1_SIMPLE
ASN1_SEQUENCE_END
IMPLEMENT_ASN1_FUNCTIONS
EC_POINT_add
i2d_PUBKEY
EC_GROUP_dup
EC_POINT_invert

SparkiDev replied:
Add a --enable-wpas-dpp (if that is the only feature that needs this) configuration option so that they have to explicitly choose the big build.

julek-wolfssl replied:
Added. Note that I implemented DPP support in https://github.com/julek-wolfssl/wolfssl/tree/openssh-fixes so as to not delay this PR.

julek-wolfssl commented on src/internal.c:
```diff
@@ -9501,7 +9503,21 @@ int CopyDecodedToX509(WOLFSSL_X509* x509, DecodedCert* dCert)
             ret = MEMORY_E;
 #if defined(OPENSSL_ALL)
-        x509->key.pubKeyOID = dCert->keyOID;
+        if (!ret) {
```
Done.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -21711,7 +21708,7 @@ int wolfSSL_i2d_DHparams(const WOLFSSL_DH *dh, unsigned char **out)
         if (ret != MP_OKAY) {
             WOLFSSL_MSG("StoreDHparams error");
-            len = 0;
+            len = WOLFSSL_FAILURE;
```
Done.

julek-wolfssl committed:
Code review changes

- Don't include `ENABLED_OPENSSLALL` with `ENABLED_WPAS`
- Return length in `wolfSSL_i2d_DHparams`
- Implement `wolfSSL_EC_POINT_mul` with independent multiplication and addition if `ECC_SHAMIR` not defined
- Implment `ASN1_SIMPLE` without `offsetof` by using a dummy struct
- Style fixes

julek-wolfssl committed:
Test fixes

- Add `parameter` to `WOLFSSL_X509_ALGOR`
- Implement `wolfSSL_ASN1_TYPE_new`, `wolfSSL_ASN1_TYPE_free`, and `wolfSSL_ASN1_TYPE_set`
- Fix leak where `pval` in `wolfSSL_X509_ALGOR_set0` was lost if `aobj` was provided

julek-wolfssl committed:
New API

Add `wc_ecc_import_point_der_ex` for correct importing DER ECC point and keep `wc_ecc_import_point_der` old functionality

julek-wolfssl committed:
Add option to enable DPP support in wpa_supplicant (note DPP not yet supported as of this commit)

julek-wolfssl committed:
wolfSSL_RSA_*_PKCS1_PSS rewrite

julek-wolfssl committed:
Revert wc_OBJ_sn2nid

julek-wolfssl commented:
Jenkins please test again

JacobBarthelmeh commented on src/ssl.c:
```diff
@@ -42163,51 +42883,25 @@ void* wolfSSL_GetDhAgreeCtx(WOLFSSL* ssl)
-                    eccEnum = ecc_sets[i].id;
-                    /* Convert enum value in ecc_curve_id to OpenSSL NID */
-                    return EccEnumToNID(eccEnum);
+            if (ln[lnlen-1] == '=') {
```
that works......for the access here since ln has also been incremented by 1 , but I suggest we avoid accessing arrays with a negative index value and we could add in an error out early case instead

JacobBarthelmeh commented on wolfssl/openssl/asn1.h:
```diff
@@ -78,7 +79,72 @@
+                                      const WOLFSSL_ASN1_ITEM *template);
+
+/* Need function declaration otherwise compiler complains */
+#define IMPLEMENT_ASN1_FUNCTIONS(type) \
```
This looks very similar to the way openssl implements functions in openssl/asn1t.h???

We don't implement functions this way in the rest of wolfSSL.

julek-wolfssl replied:
This is strictly for external use. hostapd uses this in `src/common/dpp.c` to implement `DPP_BOOTSTRAPPING_KEY`. For now this functionality is limited to only a few types but it can serve a framework for any future projects that use more complex types. I'll add a comment that this isn't for internal use.

toddouska commented:
Please resolve merge conflicts and Jacob's suggestions.

julek-wolfssl commented on wolfssl/wolfcrypt/ecc.h:
```diff
@@ -632,15 +635,23 @@ int wc_ecc_export_private_raw(ecc_key* key, byte* qx, word32* qxLen,
 int wc_ecc_export_point_der(const int curve_idx, ecc_point* point,
                             byte* out, word32* outLen);
+WOLFSSL_API
+int wc_ecc_export_point_der_compressed(const int curve_idx, ecc_point* point,
```
I would propose making `wc_ecc_export_point_der_compressed` a `WOLFSSL_LOCAL` API. What do you think?

JacobBarthelmeh replied:
Sure, having it as local is okay at this point, or having the _ex one local. The comment was mainly from a concern that if we add too many public api's that do close to the same things, that it may make it harder for users to pick which one they should call.

julek-wolfssl commented on wolfssl/wolfcrypt/ecc.h:
```diff
@@ -632,15 +635,23 @@ int wc_ecc_export_private_raw(ecc_key* key, byte* qx, word32* qxLen,
 #ifdef HAVE_ECC_KEY_IMPORT
 WOLFSSL_API
+int wc_ecc_import_point_der_ex(byte* in, word32 inLen, const int curve_idx,
+                               ecc_point* point, char shortKeySize);
```
Changed to int.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -30976,6 +31423,210 @@ WOLFSSL_ASN1_INTEGER* wolfSSL_BN_to_ASN1_INTEGER(const WOLFSSL_BIGNUM *bn, WOLFS
+                           (((byte*)src) + member->offset);
+                len += SetBitString(bit_str->length, 0, bufLenOrNull(buf, len));
+                if (buf && bit_str->data) {
+                    XMEMCPY(buf + len, bit_str->data, bit_str->length);
```
`wolfSSL_ASN1_item_i2d` first calls itself with a `NULL` `dest` parameter and then allocates the buffer based on the calculated length. This should be enough to guarantee that the length of the buffer is sufficient. If called with a buffer provided in `*dest` then the user is trusted to have provided a long enough buffer. Maybe `wolfSSL_ASN1_item_i2d` should check that the final length matches the allocated length?

JacobBarthelmeh replied:
Okay. This is another one of the cases where they blindly trust the users buffer if it has been passed in, and the function does not take in a length value.

julek-wolfssl committed:
Code review

- make `wc_ecc_export_point_der_compressed` a local function
- use `int` for `shortKeySize` in `wc_ecc_import_point_der_ex`
- check for null return value from `wolfSSL_OBJ_nid2obj` and `wolfSSL_d2i_PUBKEY`
- add comments to `ssl.c`
- check `lnlen` in `wolfSSL_OBJ_ln2nid`

julek-wolfssl committed:
Merge remote-tracking branch 'wolfSSL/master' into wpa-supplicant-openssl-compat

julek-wolfssl committed:
Fix merge issues
